content/advent-2019/go-compilers.md

```diff
@@ -0,0 +1,226 @@
+ln -s $GOFRONTEND/go gcc/go/gofrontend
+rm -rf libgo
+mkdir libgo
+for f in $GOFRONTEND/libgo/*; do ln -s $f libgo/`basename $f`; done
+./contrib/download_prerequisites
+mkdir objdir
+cd objdir
+
+# depending on your distro, you might need to re-link /bin/sh to /bin/bash instead of /bin/dash.
+
+../configure \
+    --enable-languages=c,c++,go \
+    --disable-libquadmath \
+    --disable-libquadmath-support \
+    --disable-werror \
+    --disable-multilib  \
+    --enable-gold
+make
+make install
+
+# or /usr/local/lib on 32-bit systems
+export LD_LIBRARY_PATH=/usr/local/lib64
+```
+
+To build projects using gccgo, you then invoke `go build -compiler gccgo`. However, there is a catch: gccgo builds are dynamically linked by default. You can confirm this by running `ldd` against the resulting executable. For example, running `ldd` against the binary resulting from building Caddy (v1) from source with gccgo shows:
+
+```bash
+linux-vdso.so.1 (0x00007fff98b21000)
+libgo.so.15 => /usr/local/lib64/libgo.so.15 (0x00007f00b12d7000)
+libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f00b0f39000)
+libgcc_s.so.1 => /usr/local/lib64/libgcc_s.so.1 (0x00007f00b0d21000)
+libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f00b0930000)
+/lib64/ld-linux-x86-64.so.2 (0x00007f00b2d9a000)
+libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f00b0711000)
+```
+
+We can see in the output how libgo, which is the Go stdlib as defined by the gofrontend, is dynamically linked, among other libs. The first step in our adventure to produce statically linked executable is to link libgo statically by using `-gccgoflags '-static-libgo'`. Again, building Caddy with `go build -compiler gccgo -gccgoflags '-static-libgo'` then running `ldd` shows the list of dynamically linked libraries as before sans the libgo line `libgo.so.15 => /usr/local/lib64/libgo.so.15 (0x00007f00b12d7000)`.
+
+```bash
+linux-vdso.so.1 (0x00007ffc06dfa000)
+libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7bfb121000)
+libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7bfad83000)
+libgcc_s.so.1 => /usr/local/lib64/libgcc_s.so.1 (0x00007f7bfab6b000)
+libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7bfa77a000)
+/lib64/ld-linux-x86-64.so.2 (0x00007f7bfb340000)
+```
+
+The last step is to link everything statically by using `-gccgoflags '-static'`, thus a full command would be `go build -compiler gccgo -gccgoflags '-static'`. This brings another catch: some packages have dependency on cgo, and thus fail to compile statically. For example, the `golang.org/x/sys` package limits gccgo builds to an implementation that relies on C with the build tags [here](https://github.com/golang/sys/blob/master/unix/gccgo.go) and [here](https://github.com/golang/sys/blob/master/unix/gccgo_c.c). Thus when trying to compile Caddy with `CGO_ENABLED=0 go build -compiler gccgo -tags netgo -gccgoflags '-static'` you'll get the error `undefined reference to 'gccgoRealSyscall'`.
```

Comment:
> thus a full command would be `go build -compiler gccgo -gccgoflags '-static'`

Should the full command also include `-static-libgo`?

Reply:
My understanding is the `-static` flag implies `-static-libgo`. Reading the [reference page](https://golang.org/doc/install/gccgo#Using_gccgo) again, I think it isn't 100% clear whether `-static` implies `-static-libgo` or not:

> Use the -static-libgo option to link statically against the compiled packages.
Use the -static option to do a fully static link (the default for the gc compiler).


Either way, I tried it with the inclusion of `-static-libgo` and still fails with the same error.